### ISSUE TITLE
fn:void: Absorb result of evaluated argument

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#359]
### Summary

Absorb the result of the evaluated argument.

### Signature

```xquery
fn:void(
  $input as item()*
) as empty-sequence()
```

### Motivation

Developers tend to get creative if they want to suppress the result of an expression. The reason is that there is no simply solution to do this properly. Some constructs I have seen in practice:

```xquery
let $unused := EXPRESSION
return 'ok'

EXPRESSION[position() = 10000], 'ok'

let $result := 'ok'
return if(exists(EXPRESSION)) then $result else $result
```

Cases like this are frequent in nondeterministic code. Think e.g. of side-effecting functions of the EXPath HTTP-Client and File Modules: The function results are not always relevant for the invoking application, or already known.

The function is also helpful during development and for testing code. `fn:void#1` and `fn:identity#1` can both be passed on to functions to either return or ignore the result of their arguments. The function can potentially be used to measure the runtime performance of an expression (but an implementation should not be prevented from discarding the function call if the argument expression is deterministic).

**Comment by michaelhkay** [comment]
Is there an expectation that the argument value is evaluated and the result ignored, or is the expectation that the argument is not evaluated? Your reference to side-effecting functions in EXPath suggests the former. Is there any way we can formalise this expectation, other than in non-normative notes?

**Comment by ChristianGruen** [comment]
> Is there an expectation that the argument value is evaluated and the result ignored, or is the expectation that the argument is not evaluated? Your reference to side-effecting functions in EXPath suggests the former. Is there any way we can formalise this expectation, other than in non-normative notes?

We could formalise this by stating that expressions must be evaluated if they are nondeterministic. `void(error())` should always raise an error.

To be strict, nondeterministic functions need not necessarily be side-effecting, but it would be a pretty big step to introduce side-effecting semantics into the official spec (after all, I assume we won’t introduce such functions in the standard function set too soon).

> I typically use a declared return type of `item()*` rather than `empty-sequence()` with such functions to avoid them being optimized away. But perhaps that's just an implementation detail that need not affect the spec?

I would tend to use `empty-sequence()` as the official type, as the function never returns anything. This type is also used multiple times in the [EXPath File Module Specification](http://expath.org/spec/file) and in our extensions. `none` (as used for `fn:error`) could be another option, though.

**Comment by dnovatchev** [comment]
Maybe I completely don't understand this, but it seems like a shy attempt to do lazy/suppressed evaluation.

If the argument may be evaluated, then why bother call the function at all?

**Comment by ChristianGruen** [comment]
> If the argument may be evaluated, then why bother call the function at all?

Think of e.g. the EXPath function `http:send-request` and its usage with PUT or DELETE (i.e., a method that performs updates, but won't usually return a result). A function call returns one or more items, but there may be no need to process the results any further. `fn:void` could be used to prevent the user from using `...[position() > 2]` constructs:

```xquery
http:send-request((), $url) => void()
```

Similarly, in BaseX, the results of invocations of functions like `sql:execute` or `proc:execute` may be irrelevant to the caller of the function.

**Comment by michaelhkay** [comment]
Another use case is 

`trace($x instance of xs:integer, "$x is an integer?") => void()`

where you want to output the value of some expression without actually using that value.

I currently tend to write this as

`trace((), "$x is an integer? " || $x instance of xs:integer)`

which is pretty clunky.

**Comment by ChristianGruen** [comment]
For debugging, we could even think about adding a dedicated function, such as `fn:dump`.

**Comment by benibela** [comment]
how about

    trace($x instance of xs:integer, "$x is an integer?")[0]

**Comment by ChristianGruen** [comment]
> ```
> trace($x instance of xs:integer, "$x is an integer?")[0]
> ```

Query processors may decide to rewrite `EXPR[0]` to `()`, regardless of what `EXPR` is. The same applies to `tail(trace(...))`, `trace(...)[position() >  1]` or whatever you’re inclined to try (as indicated in the initial comment of the issue).

**Comment by dnovatchev** [comment]
I have tried several times to understand the purpose of the proposed function `void()` and still can not.

Apart from the need to have one or more convincing and understandable use cases besides the definition of the function, it seems to me that the confusing thing about this function is its name.

If I understand the intent, then a better and more precise name would be:

```xq
fn:eager-and-discard($someExpression as item()*) as item()*
```
We may also need to eagerly evaluate an expression **and use its result**. This may be a new **_`fn:eager`_**, but to save keystrokes **it seems better to have this just as a keyword** that precedes the expression that we want to be evaluated no matter what:

**_`eager    Expression`_**

And if we have either a function or keyword **_`eager`_**, then it seems that nobody will need to use `fn:eager-and-discard`, just using **_`eager`_** will suffice.

Or maybe I am totally missing something? If so, please, advise.

**Comment by michaelhkay** [comment]
We've introduced an XSLT instruction saxon:do that is very similar to this, in that its purpose is to evaluate a supplied expression and discard the result, typically because the supplied expression has side-effects (but another use case is to trigger errors and catch them).

There's definitely a use case. The problem is that it's very difficult to define the semantics with any rigour.

**Comment by rhdunn** [comment]
@dnovatchev `fn:eager` has a slightly different meaning to the proposed `fn:void`, though. Specifically:
1. `fn:void` -- Evaluate the expression and ignore the return value ~ used in cases like sending HTTP requests, and saving documents in a database.
2. `fn:eager` -- Evaluate the expression eagerly at the point at which the function call is evaluated. This is akin to taking a Stream/Sequence/Linq Expression in languages like Java/Kotlin/C# and converting the result to a list -- the sequence is evaluated eagerly at that point. -- __Note:__ MarkLogic has an [xdmp:eager](https://docs.marklogic.com/xdmp:eager) function for this, which can be used in database searches.
3. `fn:lazy` -- Defer the evaluation of the expression until the point at which it is used. This is akin to using/returning a Stream/Sequence/Linq Expression and enumerating that object at the point at which it is used instead of converting it to a list. -- __Note:__ MarkLogic has an [xdmp:lazy](https://docs.marklogic.com/xdmp:lazy) function for this, which can be used in database searches.

That is, `fn:void` always returns an empty sequence, whereas `fn:eager` and `fn:lazy` would return the values like `fn:identity` does. -- The difference with `fn:identity` and `fn:eager`/`fn:lazy` is that `fn:eager`/`fn:lazy` are hints to the XPath/XQuery/XSLT processor (e.g. whether to keep/use a streaming version of an expression or convert it to a list); `fn:identity` does not/should not modify the underlying representation of the resulting value/sequence.

The point is that the function will look something like this:
```
declare function fn:void($value as item()*) as empty-sequence() {
    ()
};
```
whereas `fn:identity` returns the passed value, so would look something like this:
```
declare function fn:identity($value as item()*) as item()* {
    $value
};
```

**Comment by dnovatchev** [comment]
> The point is that the function will look something like this:
> 
> ```
> declare function fn:void($value as item()*) as empty-sequence() {
>     ()
> };
> ```
> 
> whereas `fn:identity` returns the passed value, so would look something like this:
> 
> ```
> declare function fn:identity($value as item()*) as item()* {
>     $value
> };
> ```

@rhdunn , Do I correctly understand from this statement that (you think) that **_`fn:identity`_** already has the same behavior as the proposed **_`fn:eager()`_** ?

**Comment by rhdunn** [comment]
@dnovatchev Not exactly.

In a language like Java/Kotlin/C# they would be:
```kotlin
// Do not change the value of the sequence passed to identity.
fun <T> identity(T value): T { return value; }

// Do not change the value of the sequence passed to eager.
fun <T> eager(T value): T { return value; }
// ... unless it is a lazy sequence (or linq expression), in which case evaluate it eagerly.
fun <T> eager(Sequence<T> value): List<T> { return value.toList(); }
```

Note that a processor could be more selective in how the expressions get processed. I.e. it could have two variants of functions, one returning a lazy list and another an eager list, and when processing the expression determine which variant to use/call given the use of `fn:eager` or `fn:lazy` (otherwise use some sort of internal heuristic to do this).

See also e.g. https://learn.microsoft.com/en-us/dotnet/standard/linq/deferred-execution-lazy-evaluation.

**Comment by rhdunn** [comment]
@dnovatchev The result of `fn:eager`/`fn:lazy` will *very much* be used. -- The canonical use case in MarkLogic is for database searches. E.g.:
```
(: MarkLogic determines evaluation :)
for $user in cts:search(/users, cts:true-query())
return $user/name
```
vs:
```
(: Always eager :)
for $user in xdmp:eager( cts:search(/users, cts:true-query()) )
return $user/name
```
vs
```
(: Always lazy :)
for $user in xdmp:lazy( cts:search(/users, cts:true-query()) )
return $user/name
```

**Comment by michaelhkay** [comment]
Note also that there can be different interpretations of what "eager" and "lazy" evaluation mean. Consider the expression "1 to 1000". Does eager evaluation require enumeration of the values as a list, or is an internal optimized representation of the integer range allowed? Similarly `$x/child::node()` - must the result be a free-standing list, or can it be a reference to the list of children of an XDM node within an XDM tree? Clearly these are internal implementation choices, and if we have functions like lazy() and eager() we should only prescribe externally-observable aspects of their behaviour.

**Comment by rhdunn** [comment]
@dnovatchev Note that async/await is different from eager/lazy.

An async function is one that can be (and often is) evaluated on a separate thread. It may also be a function where the result is deferrred to some point in the future, such as the result of a HTTP request. Such functions return a `Future<T>`/`Promise<T>` value.  Awaiting on such a function is equivalent to joining on the future, or calling the `.then()` method of a promise.

Lazy evaluation is different in that the result is computed on demand. It is equivalent to returning an iterator where the next method computes the next value on demand. -- As such, if the lazy sequence is reused, the values are recomputed. Requesting eager evaluation is requesting (where it makes sense for expensive operations) that the results are computed up-front, so if it is reused there isn't the additional computational cost of recalculating the value.

Note that a query processor is not required to evaluate the expressions in the order that they are written, nor how the items in the sequence are computed. This allows the processor to perform optimizations like those that @michaelhkay mentioned. It could also compute results in parallel if it had that capability (e.g. internally creating async functions for each of the let variables and awaiting at the point at which the values are requested).

**Comment by dnovatchev** [comment]
> @dnovatchev The result of `fn:eager`/`fn:lazy` will _very much_ be used. -- The canonical use case in MarkLogic is for database searches.

@rhdunn Sorry, maybe I was not clear. I said:



> As for the difference between `fn:eager` and `fn:void`, who cares that fn:eager returns a result -- it simply is not going to be used. Let us leave these details to the Optimizer.

Not that we do not care about the result of `fn:eager` in general, but that we can perfectly live having only `fn:eager` and use it instead of `fn:void`  and **only in such cases** we will not care about the result. 

Is this clear now?

**Comment by rhdunn** [comment]
@dnovatchev Thanks for the clarification.

The use cases in the original for `fn:void` show the need to exclude the result of an inner expression from the result of the outer expression. Hence:
```
let $unused := EXPRESSION
return 'ok'
```
being one of the example constructs noted by Christian seen in practice. -- Here, you don't want the result of `EXPRESSION` being added to the results of the outer FLWOR expression.

Note that I've also seen this pattern being used in production, but generally as a way of chaining expressions (e.g. when creating documents in a database).

That is:
```
1 => fn:void() , 2 => fn:eager()
```
would return `2`, *not* `(1, 2)`.

**Comment by dnovatchev** [comment]
> @dnovatchev Note that async/await is different from eager/lazy.

@rhdunn Yes, and this is exactly what I was saying (there would be different names for such functions):

> But we do not have any asynchrony defined at present, and if/when we do, there would be better names for **_`async`_** execution and **_`wait`_**.

There could be cases when the code doesn't depend on the result of evaluating Expr (which means doesn't use any values returned from this evaluation). In such cases the code execution plan could be optimized not to wait for the evaluation of Expr to be completed, but to continue immediately.

Anyway, it would be better if we use another function for such cases. Imagine that 

(1)  one does a PUT, but later in the code accesses this resource and needs its new contents. In this case there must be wait for completion of the PUT.

(2) If the PUT is aimed to be performed and its results not used, then one might want to not wait for the completion of this PUT.

It seems that an Optimizer will not be intelligent enough to detect case (1) thus separate functions for case (1) and case (2) are necessary.

**Comment by Arithmeticus** [comment]
Having read the thread, and listened to the discussion today, I agree that the rationale for the function, and its definition, needs development.

Are there non-development use-cases for `void()`?

As for development, my `void()` equivalent is to assign some operation to a variable that does not get used later. In some contexts a processor evaluates that variable, but in others it does not. When I want to force eager evaluation, I use the variable later, say in a message. When I want lazy evaluation, I comment the line out. Is `void()` mainly a convenience function for that sort of coding practice, or is it doing something more?

**Comment by rhdunn** [comment]
The way I'm defining eager vs lazy is akin to List vs Stream in Java. -- Given:
```
let $a := ( math:sin(3), math:sin(4) )
return ($a, $a)
```
1. lazy evaluation would call `math:sin` 4 times as the sequence in `$a` will be evaluated twice.
2. eager evaluation would call `math:sin` 2 times as the sequence is precomputed/evaluated once.

IIUC, @dnovatchev you (and Michael) are using eager to mean that the sequence would be evaluated at the declaration of `$a` and lazy to mean it possibly being evaluated at the return statement (i.e. the point at which it is used, not the point at which it is declared).

**Comment by michaelhkay** [comment]
Trying to expand my thoughts during today's discussion, which I was only able to articulate very briefly:

The case for this function is all about forcing evaluation of something for the sake of its side-effects.

This is precisely the issue we have with `xsl:result-document`. The way we solve it there is to allow the instruction to be evaluated only when in "final output state". Abstracting this somewhat, each expression specifies whether it evaluates its operands in final output state; if it does so (and is itself executed in final output state), the implication is that it always evaluates all its operands. In effect, in final output state there is some kind of guarantee that an expression will be evaluated exactly once. The evaluation is not necessarily "eager" in the sense that it completes before the expression returns (indeed, Saxon evaluates `xsl:result-document` asynchronously), but the important thing is that it is guaranteed to happen exactly once. So long as you're in final output state, calling a side-effectful function is reasonable clean in its effect; once you're not (for example, while evaluating a predicate), it becomes very unpredictable.

So I can see the use case for fn:void(EXPR) (although I might perhaps prefer to write it `do {EXPR}` to emphasise that it is not a real function). But I do think it needs to be accompanied by some kind of language semantics that guarantee a level of predictability, that stops people writing `let $unused := void(EXPR)` and expecting that this is enough to force EXPR to be evaluated.

**Comment by michaelhkay** [comment]
In Reece's example it's certainly possible to evaluate `math:sin()` four times, but that certainly isn't what Saxon means by lazy evaluation. Saxon will typically evaluate $a the first time it is referenced, but will then retain that value and use it the second time it is referenced. (With recent versions of Saxon, this code will start off using lazy evaluation of $a, but will quickly discover that this strategy gives no benefits, and will then switch to eager evaluation). But I really can't see any direct relevance to the proposed `fn:void()`.

**Comment by michaelhkay** [comment]
>I wonder if a function/variable annotation would be useful here. For example %side-effecting -- I.e. the value has side effects, please don't optimize the computation away.

That's the easy part. The tricky part is when someone writes `count($x / fn:void(*)) gt 10'`. Suppose the query is running against some kind of DOM that has some kind of event listener that monitors every time you touch an element node. How much optimization is this query allowed or not allowed to do? Does the usual rewrite to `exists($x / fn:void(*))[11]` have to be suppressed? What if the expression actually calls fn:void() indirectly, or dynamically?

**Comment by ChristianGruen** [comment]
I’m still trying to find an example that’s both intuitive and self-contained. Imagine you have a regex and you want to check if it’s valid. It could be done as follows:

```xquery
declare function regex:valid($regex) {
  try {
    void(matches('', $regex)),
    true()
  } catch err:FORX0002 {
    false()
  }
};

let $regex := '['
where regex:valid($regex)
return ...
```

Maybe you immediately have a better solution in mind, though? ;·)

**Comment by dnovatchev** [comment]
> The evaluation is not necessarily "eager" in the sense that it completes before the expression returns (indeed, Saxon evaluates `xsl:result-document` asynchronously), but the important thing is that it is guaranteed to happen exactly once. So long as you're in final output state, calling a side-effectful function is reasonable clean in its effect; once you're not (for example, while evaluating a predicate), it becomes very unpredictable.

@michaelhkay,

Doesn't the concept of "**_final output state_**" lose its meaning when we call `fn:transform` and use the result of the transformation including all the created _result-documents_ so far?

It seems to me this is one way of achieving eager evaluation even in ver. 3.1, or am I wrong?

People have been doing "impossible things" using this approach.

**Comment by michaelhkay** [comment]
>Doesn't the concept of "final output state" lose its meaning when we call fn:transform...?

No, I don't think so. `fn:transform` itself is side-effect free because the contents of result documents are returned as part of the `fn:transform` result (unless of course you call an extension function with side-effects, in which case you're outside the scope of the spec). Within the called transformation, the concept of "final output state" is still sufficient to provide "execute exactly once" semantics for the `xsl:result-document` instruction. It does this without constraining the order of evaluation (I'm not sure what you mean by "so far") and without forcing anything to be evaluated eagerly or lazily.

**Comment by liamquin** [comment]
In a talk at XML Prague i used xsl:message to force evaluation of file:write(), wrote a Perl script to /tmp and then executed it with a Java extension function.

This isn't optimal, and was considered extreme and obscure by many in the audience, entirely fairly.

This concept of "please evaluate this expression even though i don't use the result" feels to me similar to "unordered" expressions in XQuery (which were a mistake - it was data sources that were unordered really) and is closer to the territory of pragmas.

in XQuery you can `write let :x := unordered { _expr_ }` - maybe adding, ```
let $x := required { file:write("...") }
```
would work better than a function for this purpose?

This is separate from lazy/eager evaluation.

**Comment by michaelhkay** [comment]
Indeed, it might be worth revisiting the semantics of `ordered { EXPR }`. I don't think it has ever been well defined and I haven't come across anyone using it.

**Comment by dnovatchev** [comment]
> > The evaluation is not necessarily "eager" in the sense that it completes before the expression returns (indeed, Saxon evaluates `xsl:result-document` asynchronously), but the important thing is that it is guaranteed to happen exactly once. So long as you're in final output state, calling a side-effectful function is reasonable clean in its effect; once you're not (for example, while evaluating a predicate), it becomes very unpredictable.
> 
> > @michaelhkay,
> > 
> > Doesn't the concept of "**_final output state_**" lose its meaning when we call `fn:transform` and use the result of the transformation including all the created _result-documents_ so far?



> No, I don't think so. `fn:transform` itself is side-effect free because the contents of result documents are returned as part of the `fn:transform` result (unless of course you call an extension function with side-effects, in which case you're outside the scope of the spec). Within the called transformation, the concept of "final output state" is still sufficient to provide "execute exactly once" semantics for the `xsl:result-document` instruction. It does this without constraining the order of evaluation (I'm not sure what you mean by "so far") and without forcing anything to be evaluated eagerly or lazily.

By "use the result of the transformation including all the created _result-documents_ **so far**?" I meant that whenever we get the result from the call to `fn:transform`, we can also access all the result-documents created by the transformation passed in this same fn:transform - call.  A transformation cannot access any result-document that it creates, but if this transformation was performed within a call to `fn:transform`, then these result-documents can be accessed by the caller.

Isn't this, in fact, **eager evaluation**?

**Comment by dnovatchev** [comment]
> > > The evaluation is not necessarily "eager" in the sense that it completes before the expression returns (indeed, Saxon evaluates `xsl:result-document` asynchronously), but the important thing is that it is guaranteed to happen exactly once. So long as you're in final output state, calling a side-effectful function is reasonable clean in its effect; once you're not (for example, while evaluating a predicate), it becomes very unpredictable.
> 
> 
 > > @michaelhkay,
 > > Doesn't the concept of "**_final output state_**" lose its meaning when we call `fn:transform` and use the result of the transformation including all the created _result-documents_ so far?
> 
 > No, I don't think so. `fn:transform` itself is side-effect free because the contents of result documents are returned as part of the `fn:transform` result (unless of course you call an extension function with side-effects, in which case you're outside the scope of the spec). Within the called transformation, the concept of "final output state" is still sufficient to provide "execute exactly once" semantics for the `xsl:result-document` instruction. It does this without constraining the order of evaluation (I'm not sure what you mean by "so far") and without forcing anything to be evaluated eagerly or lazily.
> 
 By "use the result of the transformation including all the created _result-documents_ **so far**?" I meant that whenever we get the result from the call to `fn:transform`, we can also access all the result-documents created by the transformation passed in this same fn:transform - call. A transformation cannot access any result-document that it creates, but if this transformation was performed within a call to `fn:transform`, then these result-documents can be accessed by the caller.
 
 Isn't this, in fact, **eager evaluation**?

**Comment by liamquin** [comment]
On Tue, 2023-04-25 at 15:10 -0700, Michael Kay wrote:
> Indeed, it might be worth revisiting the semantics of `ordered { EXPR
> }`. I don't think it has ever been well defined and I haven't come
> across anyone using it.
> 

The people who wanted it for XQuery were relational database vendors.
The reason they gave was that they keep a shared cursor for a large
table, and return results from wherever it happens to be.

In the end, though, they implemented XQuery only within a BLOB or
equivalent, not over views of tables, so i am not at all sure anyone
implemented it in the end.

This reasoning, though, is why i say it was badly designed - what was
wanted was to be able to say a specific data source was unordered.

-- 
Liam Quin, https://www.delightfulcomputing.com/
Available for XML/Document/Information Architecture/XSLT/
XSL/XQuery/Web/Text Processing/A11Y training, work & consulting.
Barefoot Web-slave, antique illustrations:  http://www.fromoldbooks.org

**Comment by ChristianGruen** [comment]
> The people who wanted it for XQuery were relational database vendors. The reason they gave was that they keep a shared cursor for a large table, and return results from wherever it happens to be. In the end, though, they implemented XQuery only within a BLOB or equivalent, not over views of tables, so i am not at all sure anyone implemented it in the end.

That's interesting. – We ignore `ordered` and `unordered` expressions, and I don't remember to have seen code in which it was used.

**Comment by michaelhkay** [comment]
>Isn't this, in fact, eager evaluation?

I can't see why. All the caller knows is that the result document is available at the time they access it. They have no way of knowing when it was actually evaluated.

In the current Saxon implementation, maps and arrays (unlike sequences) are always "grounded", meaning that the contents are always present in memory and require no further evaluation. Because the result of fn:transform is a map, this will be true of the result documents that are returned. But this is simply a limitation of the current implementation, there is no intrinsic reason why maps and arrays shouldn't contain entries that are evaluated on first reference.

**Comment by michaelhkay** [comment]
`ordered` and `unordered` are a classic case of a feature added to the language because implementors thought they could take advantage of it, not because users thought they had a need for it. Such features are almost always a mistake and end up being unused and probably unimplemented. A mistake we must learn from.

**Comment by ChristianGruen** [comment]
Here’s a related example from @line-o on Slack that leads to different results with different processors (and, possibly, versions):

```xquery
try {
  let $_ := matches("", "\")
  return "All fine."
} catch * {
  $err:description
}
```

**Comment by dnovatchev** [comment]
One way to implement `fn:eager` is to have a standard (or internal) hashing function:

```xq
$fn:hash($arg as xs:string) { 
(:  returns the hash, for example Blake3 (https://en.wikipedia.org/wiki/BLAKE_(hash_function))  of the sting  :)
} as xs:string 
.   .   .   .   .   .   .
```

Then one way to define **`fn:eager`** is:

```xq
fn:eager($expr as item()*) as xs:integer
{
   if(hash(serialize($expr) ne "1234567890")
     then 1
     else error()
}
```
And then in the user's code:

```xq
   if(fn:eager(Expr) gt 0)
      then (: Do all the work :)
      else ()
```

**Comment by michaelhkay** [comment]
I'm really struggling to know what you're trying to achieve.

You've got an expression EXPR that you want evaluated for its side effects.

The problem is that in some contexts, it won't be evaluated because the result isn't needed. For example, if you write

```
let $unused := EXPR
return 3
```
then the optimizer might decide it doesn't need to evaluate EXPR because the result isn't going to be used.

So you change it to

```
let $unused := eager(EXPR)
return 3
```
and exactly the same thing happens. If the result wasn't needed before, then it still isn't needed now.

How is this function supposed to help resolve this situation?

**Comment by ChristianGruen** [comment]
If at all, we could possibly improve things if we tweak the Errors & Optimization section of the spec to enforce fewer rewritings for nondeterministic expressions. If someone writes…

```xquery
let $unused := error()
return 3
```

…we can safely assume that the user expects the unreferenced code to be executed. Same for `let $xyz := file:write('x.txt', $text)` and other nondeterministic/side-effecting expressions.

If code is deterministic, though, it makes no difference if the expression is ignored… Unless it may raise errors at compile or runtime. It can happen that an optimizer replaces a deterministic function – `fn:matches("", "\")` – with a nondeterministic one – `fn:error`.

**Comment by ChristianGruen** [comment]
> No optimization to be performed in here.

Personally, I’d be happy to close the issue. There are so many cases in which I couldn't even tell if a specific compiler action is an optimization or just a rewriting for another internal representation. It’s fairly easy to think about this if one starts from scratch, and if it's clearly defined in the spec. It’s close to impossible if one wants to catch up on this.

And I still think that a void function and eager/lazy evaluation are different things.

**Comment by michaelhkay** [comment]
"we can safely assume that the user expects the unreferenced code to be executed"

No we can't. It's much more likely that it's dead code left around from uncompleted changes. And in practice the situation is often much more subtle, e.g. the variable is actually referenced somewhere, but in a branch of a conditional that isn't actually executed. Or the variable is referenced but we know from type analysis that it's going to be an empty sequence so we don't need to know anything about the content.

The problem with any construct that attempts to force evaluation of its operands is that you also have to force evaluation of that construct, and so on, all the way up to the top level. That's essentially how we solve the problem with xsl:result-document: it can only be used in a context ("final output state") where execution is linear and predictable, all the way up to the top.

**Comment by ChristianGruen** [comment]
> No we can't. It's much more likely that it's dead code left around from uncompleted changes.

Sorry, I think my wording was too unspecific. I wanted to point out it's (unfortunately) a common pattern for people to use unreferenced variables to try to enforce an evaluation order, in particular if the expressions are nondeterministic, and in particular if people are used imperative languages.

Maybe we can turn things around and try to find out first what would be a better writing with the existing language syntax for examples like the ones above?

```xquery 
declare function regex:valid($regex) {
  try {
    void(matches('', $regex)),
    true()
  } catch err:FORX0002 {
    false()
  }
};

try {
  let $_ := matches("", "\")
  return "All fine."
} catch * {
  $err:description
}
```

Which rewritten version would yield the same result for all processors?

**Comment by line-o** [comment]
I do like the concept of `fn:void` to call functions that return something where this value is not needed in the context.
Is there a way to achieve this without this.
@ChristianGruen first example is also my preferred way of application:

```
void(my:f()),
true()
```

With processors optimising expression evaluation like `head((true(), my:f()))` or `(true(), my:f())[1]` I wonder how to achieve this. Maybe I am overseeing a _simple_ construct to call a function and discard its return value. I do think a general purpose language should offer this.

**Comment by liamquin** [comment]
orthogonality was a major design principle for XQuery and XPath. The idea of a function whose arguments are evaluated differently seems very much against the spirit of the language.

i agree with @ndw of course (since i've already suggested a language construct), e.g.
```
    required {  enclosedExpr }
```
by analogy with ordered / unordered expressions. The semantics would be that the processor must evaluate the expression in any context where an expression would normally be evaluated without optimization - e.g. in
( a, b, c, required { d } )
where there's a non-empty a element in the current context, it's not evaluated, but for
let $a := required { call-home() } return 'done'
it would be evaluated.

The question is, how easy is this (1) to specify formally, and (2) to implement?

This is nothing to do with eager or lazy; it's to do with forcing evaluation of an expression that the processor doesn't know might have side-effects.

Of course, exactly at what point it's evaluated will depend on the optimizer, so to force execution order you still have to arrange a dependency, e.g. ```
  let $result := (required { $a }, required { $b})[1]
  return $result
  ```
 wouldn't normally do that, the expressions could be evaluated in either order, maybe even in separate threads, and in any case only $a is needed, if it returns non-false. 
 
 However, required { ($a, $b) } [1] would force execution of both. I think? Coffee hasn't worked yet.
 
 I don't think it's about discarding the return value in the side-effect case, but about ensuring the function is called.

**Comment by michaelhkay** [comment]
There's no point in having a construct

`required {  enclosedExpr }`

if it's buried five-layers deep in a function called by `x` which is called by `y` which is called by `z` which is called like

`(f(), g(), h(), z())[1]`

There's no way we can make a processor call z() so that it can call y() so that it call x() because x contains something that is "required" for its side-effects.

To do this, at the very least, you would have to mark the functions all the way up as %side-effecting.

**Comment by ChristianGruen** [comment]
> The idea of a function whose arguments are evaluated differently seems very much against the spirit of the language.

@liamquin I agree. It shouldn’t be the function that mandates how the argument is treated. The only job of `fn:void` should be to discard the result and return an empty sequence.

---

If we decided to fix common principles for eager/lazy processing (I hope not), I would be surprised if the concept was adopted by anyone who’s not directly involved in the discussion. For the last hour, I’ve been running the queries from the comments above, and similar ones, with 5 XQuery processors (Zorba, Xmlprime, eXist, Saxon, BaseX). I am surprised at how differently simple queries are evaluated, with most results appearing to be correct. To me, this indicates that the existing compilation strategies are just too diverse that we could find a reasonable compromise that suits the majority of the available implementations.

**Comment by dnovatchev** [comment]
> One way to implement `fn:eager` is to have a standard (or internal) hashing function:
> 
> ```xq
> $fn:hash($arg as xs:string) { 
> (:  returns the hash, for example Blake3 (https://en.wikipedia.org/wiki/BLAKE_(hash_function))  of the string  :)
> } as xs:string 
> .   .   .   .   .   .   .
> ```
> 
> Then one way to define **`fn:eager`** is:
> 
> ```xq
> fn:eager($expr as item()*) as xs:integer
> {
>    if(hash(serialize($expr) ne "1234567890")
>      then 1
>      else error()
> }
> ```
> 
> And then in the user's code:
> 
> ```xq
>    if(fn:eager(Expr) gt 0)
>       then (: Do all the work :)
>       else ()
> ```

Folks, Nobody commented on this (copying the whole comment). That means that either it is correct (and you agree), or more probably that you somehow missed it.

I am rereading it and don't find any obvious flaw. So, please, what do you think?

**Comment by dnovatchev** [comment]
> There's no point in having a construct
> 
> `required { enclosedExpr }`
> 
> if it's buried five-layers deep in a function called by `x` which is called by `y` which is called by `z` which is called like
> 
> `(f(), g(), h(), z())[1]`
> 
> There's no way we can make a processor call z() so that it can call y() so that it call x() because x contains something that is "required" for its side-effects.
> 
> To do this, at the very least, you would have to mark the functions all the way up as %side-effecting.

@michaelhkay I think no one wants this, but what we are discussing as the proposed new feature is just **if** the function that immediately contains 

```xq
required {  enclosedExpr }
```

is being evaluated, **then**  `enclosedExpr` should be evaluated.

**Comment by ChristianGruen** [comment]
> Folks, Nobody commented on this (copying the whole comment). That means that either it is correct (and you agree), or more probably that you somehow missed it.

@dnovatchev I didn’t comment on it here, as the title of the issue is "Absorb result of evaluated argument", and as I couldn’t see how you are going to solve the challenges that I have described in the initial comment of this issue. I cannot see either how it could be used to rewrite the try/catch constructs in my recent comment.

I’ll create a new issue on eager and lazy evaluation, in which I’ll comment your hashing code. We can then use this issue to try to solve the initial problem, or just close it if we feel it’s not relevant or too difficult to solve.

**Comment by michaelhkay** [comment]
>Folks, Nobody commented on this

It's been a busy thread and some of us have other work to do. My comment was about the need for the function, not your suggested implementation. I don't see why you're concerned about implementation when (a) the need for the function hasn't been agreed, (b) implementation is entirely a matter for the implementor, and (c) implementation is hardly likely to be a problem.

**Comment by dnovatchev** [comment]
> > Folks, Nobody commented on this
> 
> It's been a busy thread and some of us have other work to do. My comment was about the need for the function, not your suggested implementation. I don't see why you're concerned about implementation when (a) the need for the function hasn't been agreed, (b) implementation is entirely a matter for the implementor, and (c) implementation is hardly likely to be a problem.

@michaelhkay Re: the need for the function:

I thought this was obvious. Using **`fn:eager`** neatly solves the initial issue for **`fn:void`** that was raised by @ChristianGruen. The latter is a special case of using `fn:eager` in which the provided expression does not produce a result.

What is most important now is if you agree or disagree with all 4 or any of the  statements listed in the comment.

Why? because having reached the state when these 4 statements are true, we have made a good progress on the issue.

And, @michaelhkay, this is in response to your statements that to make this (eager evaluation) possible, we would need some special definitions or even changes in the DM, that would be risky to make. I have shown that it is possible to achieve **_guaranteed_** eager evaluation without touching the DM, just using regular functions.

Any thoughts on the 4 statements?

**Comment by ChristianGruen** [comment]
> I thought this was obvious. Using **`fn:eager`** neatly solves the initial issue for **`fn:void`** that was raised by @ChristianGruen.

Please show us how. Feel free to rewrite my example to illustrate it:

```xquery
try {
  void(matches('', $regex)),
  true()
} catch err:FORX0002 {
  false()
}
```

**Comment by dnovatchev** [comment]
> > I thought this was obvious. Using **`fn:eager`** neatly solves the initial issue for **`fn:void`** that was raised by @ChristianGruen.
> 
> Please show us how. Feel free to rewrite my example to illustrate it:
> 
> ```xq
> try {
>   void(matches('', $regex)),
>   true()
> } catch err:FORX0002 {
>   false()
> }
> ```

Sorry, but I don't understand what this code does and proves. Could you, please, elaborate? What is the problem?

**Comment by michaelhkay** [comment]
I think that 2 use cases for fn:void have now been put forward:

(a) forcing execution of an expression that has side-effects. I have argued that it's not sufficient on its own to satisfy that requirement; this would need a much more complete semantics for order-of-execution etc.

(b) "teasing out errors" in an expression; that is, establishing that an expression will not fail when evaluated, in a situation where the actual result is not needed. It's wrong to suggest that the only way to "tease out errors" is by eager evaluation of the expression. For example, an implementation might know that adding two decimals can never fail (even though it might fail with an overflow on other implementations). Similarly, it might be able to work out from static type inferencing that an expression of the form `($a gt $b)` will never fail. Or it might know that name() is unfailing if it knows statically that the context item is a node. Whether or not an expression is unfailing and side-effect-free is a property that optimizers can make inferences about, and they should be allowed to use that knowledge freely.

I think that the second use case is probably strong enough on its own to justify the function, and I think the semantics can be described fairly clearly, in language that's at least as good as the language we use for "errors and optimization". The function will probably end up being used for both use cases, even though it's not a total solution to the first one.

The function seems to meet this requirement and I fail to see how the proposed `fn:eager()` functions adds any further value.

**Comment by dnovatchev** [comment]
> I think that 2 use cases for fn:void have now been put forward:
> 
> (a) forcing execution of an expression that has side-effects. I have argued that it's not sufficient on its own to satisfy that requirement; this would need a much more complete semantics for order-of-execution etc.
> 

I actually provided specific code that guarantees that an Expr will be evaluated. The requirement above as stated doesn't contain any order-of execution expectations, and the code I provided (in the "then" clause) will need to be evaluated only after Expr is evaluated. All this code is not based on or need any such additional complexities.

> (b) "teasing out errors" in an expression; that is, establishing that an expression will not fail when evaluated, in a situation where the actual result is not needed. It's wrong to suggest that the only way to "tease out errors" is by eager evaluation of the expression. For example, an implementation might know that adding two decimals can never fail (even though it might fail with an overflow on other implementations). Similarly, it might be able to work out from static type inferencing that an expression of the form `($a gt $b)` will never fail. Or it might know that name() is unfailing if it knows statically that the context item is a node. Whether or not an expression is unfailing and side-effect-free is a property that optimizers can make inferences about, and they should be allowed to use that knowledge freely.
> 
> I think that the second use case is probably strong enough on its own to justify the function, and I think the semantics can be described fairly clearly, in language that's at least as good as the language we use for "errors and optimization". The function will probably end up being used for both use cases, even though it's not a total solution to the first one.
> 

I believe that trying to "swallow" errors is bad practice in 99% of all cases,  and the an example for the remaining 1% is code that we want "never to fail", like a Logging component that must continue to function even after some exception in its execution -  thus this is not a representative use case for me. Other participants in the group expressed a similar opinion in our April 24th meeting.

What am I missing?

> The function seems to meet this requirement and I fail to see how the proposed `fn:eager()` functions adds any further value.

**Comment by ChristianGruen** [comment]
> Sorry, but I don't understand what this code does and proves. Could you, please, elaborate? What is the problem?

(clarified in the chat)

**Comment by michaelhkay** [comment]
>I actually provided specific code that guarantees that an Expr will be evaluated. 

I don't think it does so at all. Given an expression `(A, B, C)[1]` it is entirely likely that C will not be evaluated, and wrapping it as `(A, B, eager(C))[1]` will not change this. You could try to get around this by writing `eager((A, B, C)[1])`, or perhaps as (eager((A, B, C)))[1]` but that suffers the same difficulty one step removed. The only way around it is to force eager evaluation all the way up to the top of the execution tree.

I also think you're confusing "eager evaluation" in the sense of when something is evaluated (specifically, before the containing expression returns) with "complete evaluation" (which means that the expression is fully evaluated at some time, but not necessarily synchronously). These are different things, and if you want to persuade us that the function is going to be useful you need to be much more specific about its precise semantics.

**Comment by dnovatchev** [comment]
> I also think you're confusing "eager evaluation" in the sense of when something is evaluated (specifically, before the containing expression returns) with "complete evaluation" (which means that the expression is fully evaluated at some time, but not necessarily synchronously). These are different things, and if you want to persuade us that the function is going to be useful you need to be much more specific about its precise semantics.

Yes, I think Norm (@ndw) proposed a better name: `required` .



> > I actually provided specific code that guarantees that an Expr will be evaluated.
> 
> I don't think it does so at all. Given an expression `(A, B, C)[1]` it is entirely likely that C will not be evaluated, and wrapping it as `(A, B, eager(C))[1]` will not change this. You could try to get around this by writing `eager((A, B, C)[1])`, or perhaps as `(eager((A, B, C)))[1]` but that suffers the same difficulty one step removed. The only way around it is to force eager evaluation all the way up to the top of the execution tree.

I already replied to a similar comment earlier in this thread:



> > There's no point in having a construct
> > `required { enclosedExpr }`
> > if it's buried five-layers deep in a function called by `x` which is called by `y` which is called by `z` which is called like
> > `(f(), g(), h(), z())[1]`
> > There's no way we can make a processor call z() so that it can call y() so that it call x() because x contains something that is "required" for its side-effects.
> > To do this, at the very least, you would have to mark the functions all the way up as %side-effecting.
> 
> @michaelhkay I think no one wants this, but what we are discussing as the proposed new feature is just **if** the function/expression that immediately contains
> 
> ```xq
> required {  enclosedExpr }
> ```
> 
> is being evaluated, **then** `enclosedExpr` should be evaluated.

**Comment by ChristianGruen** [comment]
In the initial comment, I wrote:

> The function is also helpful during development and for testing code. `fn:void#1` and `fn:identity#1` can both be passed on to functions to either return or ignore the result of their arguments.

The function could be used as default for the `$fallback` parameter of [`map:get`](https://qt4cg.org/specifications/xpath-functions-40/Overview.html#func-map-get).

**Comment by dnovatchev** [comment]
It took me 20 minutes just to brush through this whole comment thread.

The result: I am not convinced at all that any two of the commentors (or the author) have common understanding of what the function is supposed to do.

Thus, **based on this huge and obvious divergence of opinions and understanding, please, let us not bring this mess and confusion into the spec**.

Based on my personal understanding (which may or may not be what the author originally thought was being suggested) of this proposal, I would prefer names such as:

- **_fn:action_**()

- **_fn:trigger-action_**()

- **_fn:trigger_**()

**Comment by dnovatchev** [comment]
Not convinced at all that we have common understanding what this function is even supposed to do. See my last comment right above.

I think it would be good not to have this in the Specification, until common understanding is reached.

Someone at least should be responsible about a minimal quality bar, or otherwise let us not lie to ourselves that we are doing useful work.

Tired of seeing this again and again ... 😞

**Comment by ChristianGruen** [comment]
> Tired of seeing this again and again ... 😞

I would really be thankful if you could give feedback on open pull requests. The PR for this issue has been open for about a month, and I’ve asked for opinions more than once.

If we keep reopening issues after finally and jointly accepting them in the meetings, we won’t finalize the spec any time soon.

**Comment by michaelhkay** [comment]
The issue thread explored a wide variety of options before converging on a proposed function whose specification is very simple. The fact that it's hard to follow the process by which we got there, and that lots of people had lots of ideas before we converged, doesn't really affect whether or not the function that we finally settled on is useful or not. I don't think the grounds you have given for reopening the issue are legitimate. If you have issues with the function that we added to the spec yesterday, it would be much better to raise your specific concerns in a new issue, rather than re-opening a thread which as you rightly say explored a wide variety of requirements and ideas.

**Comment by ChristianGruen** [comment]
Closing this in favor of #359.
